omniqueue/src/backends/memory_queue.rs

```diff
@@ -90,22 +91,60 @@ pub struct MemoryQueueConsumer {
     tx: broadcast::Sender<Vec<u8>>,
 }
 
-#[async_trait]
-impl QueueConsumer for MemoryQueueConsumer {
-    type Payload = Vec<u8>;
-
-    async fn receive(&mut self) -> Result<Delivery, QueueError> {
-        let payload = self.rx.recv().await.map_err(QueueError::generic)?;
-
-        Ok(Delivery {
+impl MemoryQueueConsumer {
+    fn wrap_payload(&self, payload: Vec<u8>) -> Delivery {
+        Delivery {
             payload: Some(payload.clone()),
             decoders: self.registry.clone(),
             acker: Box::new(MemoryQueueAcker {
                 tx: self.tx.clone(),
                 payload_copy: Some(payload),
                 alredy_acked_or_nacked: false,
             }),
-        })
+        }
+    }
+}
+
+#[async_trait]
+impl QueueConsumer for MemoryQueueConsumer {
+    type Payload = Vec<u8>;
+
+    async fn receive(&mut self) -> Result<Delivery, QueueError> {
+        let payload = self.rx.recv().await.map_err(QueueError::generic)?;
+        Ok(self.wrap_payload(payload))
+    }
+
+    async fn receive_all(
+        &mut self,
+        max_messages: usize,
+        deadline: Duration,
+    ) -> Result<Vec<Delivery>, QueueError> {
+        let mut out = Vec::with_capacity(max_messages);
+
+        // Await at least one delivery before starting the clock
+        let msg = self.rx.recv().await;
+        let delivery = msg
+            .map(|payload| self.wrap_payload(payload))
+            .map_err(QueueError::generic)?;
+        out.push(delivery);
+
+        let mut interval = tokio::time::interval(deadline);
+        // Skip the first tick which is instantaneous
+        interval.tick().await;
+        loop {
+            tokio::select! {
+                _ = interval.tick() => break,
+                msg = self.rx.recv() => {
+                    let delivery = msg
+                        .map(|payload| self.wrap_payload(payload)).map_err(QueueError::generic)?;
+                    out.push(delivery);
+                    if out.len() >= max_messages {
+                        break;
+                    }
+                }
+            }
```

Comment:
FWIW, I've had to implement something similar with in-memory queues before. I think you could simplify this greatly by using [tokio::time::timeout](https://docs.rs/tokio/latest/tokio/time/fn.timeout.html) and [try_recv()](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Receiver.html#method.try_recv)

```
let fut = async {
     // Always block for the first message
     let msgs = vec![self.rx.recv().await?];

     // Use try_recv() because it only returns values immediately available,
     // instead of blocking
     while let Ok(val) = self.rx.try_recv() {
          msgs.push(val);
     }

     msgs
};

tokio::time::timeout(fut, timeout)
```

You can't copy this verbatim, obviously, but it's more intuitive than using `select!`, `loop`, etc.

Reply:
I did finally get this to work via a combination of `tokio::time::timeout` for the initial item, then looping for the remaining time in the deadline using [`now_or_never`](https://docs.rs/futures-util/latest/futures_util/future/trait.FutureExt.html#method.now_or_never) which is a new one for me.

Once I got it working for rabbit, I followed the same pattern for the in-memory backend with `try_recv`.